pkg/neotest/signer.go

```diff
@@ -0,0 +1,101 @@
+package neotest
+
+import (
+	"bytes"
+
+	"github.com/nspcc-dev/neo-go/pkg/config/netmode"
+	"github.com/nspcc-dev/neo-go/pkg/core/transaction"
+	"github.com/nspcc-dev/neo-go/pkg/crypto/hash"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neo-go/pkg/vm/opcode"
+	"github.com/nspcc-dev/neo-go/pkg/wallet"
+)
+
+// Signer is a generic interface which can be either simple- or multi-signature signer.
+type Signer interface {
+	// ScriptHash returns signer script hash.
+	Script() []byte
+	// Script returns signer verification script.
+	ScriptHash() util.Uint160
+	// SignHashable returns invocation script for signing an item.
+	SignHashable(uint32, hash.Hashable) []byte
+	// SignTx signs a transaction.
+	SignTx(netmode.Magic, *transaction.Transaction) error
+}
+
+// signer represents simple-signature signer.
+type signer wallet.Account
+
+// multiSigner represents single multi-signature signer consisting of provided accounts.
+type multiSigner []*wallet.Account
+
+// NewSingleSigner returns multi-signature signer for the provided account.
+// It must contain exactly as many accounts as needed to sign the script.
+func NewSingleSigner(acc *wallet.Account) Signer {
+	return (*signer)(acc)
+}
+
+// Script implements Signer interface.
+func (s *signer) Script() []byte {
+	return (*wallet.Account)(s).Contract.Script
+}
+
+// ScriptHash implements Signer interface.
+func (s *signer) ScriptHash() util.Uint160 {
+	return (*wallet.Account)(s).Contract.ScriptHash()
+}
+
+// SignHashable implements Signer interface.
+func (s *signer) SignHashable(magic uint32, item hash.Hashable) []byte {
+	return append([]byte{byte(opcode.PUSHDATA1), 64},
+		(*wallet.Account)(s).PrivateKey().SignHashable(magic, item)...)
+}
+
+// SignTx implements Signer interface.
+func (s *signer) SignTx(magic netmode.Magic, tx *transaction.Transaction) error {
+	return (*wallet.Account)(s).SignTx(magic, tx)
+}
+
+// NewMultiSigner returns multi-signature signer for the provided account.
+// It must contain exactly as many accounts as needed to sign the script.
+func NewMultiSigner(accs ...*wallet.Account) Signer {
```

Comment:
Imagine if one of `accs` is a contract-based account. Then it shouldn't be possible to construct multisignature account from the provided `accs`.

Reply:
Added a check in constructor.